packages/catalog-process/src/services/readModelService.ts

```diff
@@ -0,0 +1,289 @@
+import { z } from "zod";
+import {
+  catalogItem,
+  CatalogItem,
+  Document,
+  persistentAgreement,
+  PersistentAgreement,
+  PersistentAgreementState,
+  DescriptorState,
+} from "pagopa-interop-models";
+import { match } from "ts-pattern";
+import { AuthData } from "pagopa-interop-commons";
+
+import {
+  Consumer,
+  ListResult,
+  consumer,
+  emptyListResult,
+} from "../model/domain/models.js";
+import { config } from "../utilities/config.js";
+
+const client = new MongoClient(config.mongoUri);
+
+const db = client.db("readmodel");
+const catalog = db.collection("eservices");
+const agreements = db.collection("agreements");
+
+function arrayToFilter<T, F extends object>(
+  array: T[],
+  f: (array: T[]) => F
+): F | undefined {
+  return array.length > 0 ? f(array) : undefined;
+}
+
+export const readModelGateway = {
+  async getCatalogItems(
+    authData: AuthData,
+    eservicesIds: string[],
+    producersIds: string[],
+    states: DescriptorState[],
+    agreementStates: PersistentAgreementState[],
+    offset: number,
+    limit: number,
+    name?: { value: string; exactMatch: boolean }
+  ): Promise<ListResult<CatalogItem>> {
+    const ids = await match(agreementStates.length)
+      .with(0, () => eservicesIds)
+      .otherwise(async () =>
+        (
+          await this.listAgreements(
+            eservicesIds,
+            [authData.organizationId],
+            [],
+            agreementStates
+          )
+        ).map((a) => a.eserviceId)
+      );
+
+    if (agreementStates.length > 0 && ids.length === 0) {
+      return emptyListResult;
+    }
+
+    const nameFilter = name
+      ? {
+          "data.name": {
+            $regex: name.exactMatch ? `^${name.value}$$` : name.value,
+            $options: "i",
+          },
+        }
+      : {};
+
+    const aggregationPipeline = [
+      {
+        $match: {
+          ...nameFilter,
+          ...arrayToFilter(states, (states) => ({
+            "data.descriptors": { $elemMatch: { state: { $in: states } } },
+          })),
+          ...arrayToFilter(ids, (ids) => ({ "data.id": { $in: ids } })),
+          ...arrayToFilter(producersIds, (producersIds) => ({
+            "data.producerId": { $in: producersIds },
+          })),
+        },
+      },
+      {
+        $project: {
+          data: 1,
+          computedColumn: { $toLower: ["$data.name"] },
+        },
+      },
+      {
+        $sort: { computedColumn: 1 },
+      },
+    ];
+
+    const data = await catalog
+      .aggregate([...aggregationPipeline, { $skip: offset }, { $limit: limit }])
+      .toArray();
+
+    const result = z.array(catalogItem).safeParse(data.map((d) => d.data));
+    return result.success
+      ? {
+          results: result.data,
+          totalCount: (
+            await catalog
+              .aggregate([...aggregationPipeline, { $count: "count" }])
+              .toArray()
+          )[0].count as number,
+        }
+      : emptyListResult;
```

Comment:
maybe we want to log an error here?

(also maybe we could parse the `count` with zod instead of using `as number`, and log an error in that case as well, though I guess that's unlikely to go wrong)